### ISSUE TITLE
Split Sweden to bidding zones

### ISSUE THREAD

**Issue by tmslaine** [electricitymap/electricitymap-contrib#1397]
This is not exactly compliant with the 'keep calculations to minimum' policy, but it might be possible to show a generation mix for Swedish bidding zones.

- Statnett/Svenska kraftnät have real-time flow data
- Statnett/Svenska kraftnät have a rough fuel mix for all of Sweden
- ENTSO-E has wind generation for Swedish bidding zones
- ENTSO-E/Svenska kraftnät have consumption for the bidding zones

Moreover, based on statistics by Svenska kraftnät at https://www.svk.se/aktorsportalen/elmarknad/statistik/ we can deduce:
- All nuclear capacity is located in the bidding zone SE3
- Over 98% of annual non-wind power generation in SE1 comes from hydro
- Over 97% of annual non-wind power generation in SE2 comes from hydro
- SE3 has annually 5 to 9 times more hydro generation than SE4 (on average 2010-2016 7.4 times)

Then to calculating hydro and thermal generation for each zone:
- Consumption and flow data allow calculation of total generation for each bidding zone
- We can subtract wind generation(ENTSO-E, all areas) and nuclear generation(statnett/SE3) from the total generation figures to get (hydro+unknown) generation for all areas
- By assuming that hydro makes up let's say 98% and 97% of (hydro+unknown) generation in SE1, and SE2, respectively, we have the fuel mix down for SE1 and SE2
- Subtracting SE1 and SE2 generation from the Statnett generation figures for all of Sweden, we have an estimate for the fuel mix of (SE3+SE4) and the total generation for both SE3 and SE4

The remaining problem is to solve the hydro-to-unknown ratios for SE3 and SE4:
- We could fix that something like 12% of hydro generation in (SE3+SE4) occurs in SE4, or use some MW lower/upper bounds for hydro/thermal to get the fuel mix for SE3 and SE4
- Alternatively, we could lump SE3 and SE4 together

**Comment by maxbellec** [comment]
This is awesome, thanks a lot for this. We really want to push regional consumptions as much as possible, even if it means a bit more calculations.

I'll submit a PR soon. ENTSOE publishes onshore wind per zone, but do you know if SE has any offshore wind that might not be published?

Notes for me: 
- [x] OK for exchanges, only a small modification in `statnett.py`
- [x] already done for all SE production
- [ ] ENTSOE wind production per zones: offshore wind?

**Comment by tmslaine** [comment]
Some comparison from last year between what ENTSO-E reports as onshore and what the local TSO Svenska kraftnät has as total wind production for the bidding zones, seems to be pretty much one-to-one:
![se1](https://user-images.githubusercontent.com/25774514/39880254-e70f987a-5485-11e8-9f38-312663ed530c.png)
![se2](https://user-images.githubusercontent.com/25774514/39880255-e72d15ee-5485-11e8-951c-fdc04782e07b.png)
![se3](https://user-images.githubusercontent.com/25774514/39880257-e747a38c-5485-11e8-87d9-3c4c6e8e8524.png)
![se4](https://user-images.githubusercontent.com/25774514/39880258-e76327ec-5485-11e8-9f8d-2c3d74d41991.png)

**Comment by maxbellec** [comment]
great, thanks for checking this. I hope to publish the PR tomorrow or early next week

**Comment by maxbellec** [comment]
After looking a bit more in details and almost implementing it, I think we can split all zones. There's not a lot of thermal anyway, so we won't be far from the truth by putting some of the remaining unkown from SE3 + SE4 in both of them. It'd be nice to still have a reasonable rule.
If we could know the capacities or annual production of thermal plants in SE3 and SE4 we'd get a good ratio.

**Comment by brunolajoie** [comment]
This interactive [map](https://resourcewatch.org/data/explore?zoom=4&lat=62.23323453468927&lng=33.45233917236329&basemap=dark&labels=light&layers=%255B%257B%2522dataset%2522%253A%2522a86d906d-9862-4783-9e30-cdb68cd808b8%2522%252C%2522opacity%2522%253A1%252C%2522visible%2522%253Atrue%252C%2522layer%2522%253A%25222a694289-fec9-4bfe-a6d2-56c3864ec349%2522%257D%255D&page=1&sort=most-viewed&sortDirection=-1) of power plant might help you check where are the power plants in each bidding zones

**Comment by maxbellec** [comment]
Looking at this power plant database : http://datasets.wri.org/dataset/globalpowerplantdatabase
there's only 12 non hydro, wind, nuclear power plants in Sweden. Let's just take as a proportion the estimated generation in SE3 vs estimated generation in SE4. I'll look tomorrow which power plants correspond to which zones

![screenshot at 2018-05-10 23-23-42](https://user-images.githubusercontent.com/7253902/39895117-63b93584-54a9-11e8-81a7-2928d1785247.png)

**Comment by maxbellec** [comment]
(BTW, @alixunderplatz I'm sure you're going to like this thread)

**Comment by maxbellec** [comment]
All but one power plant (that's in SE-SE4) have a 57.7 < latitude < 59.9, so are in SE-SE3. The power plant in SE-SE4 has a very high capacity of 400MW but an annual estimated generation of 135GWh, which corresponds to an average load factor of 3%.
This annual estimate generation is around 1% of SE's "unknown" generation, so if these numbers are True, we'd put 1% of unknown in SE-SE4 and the 99% remaining in SE-SE3. But I have some doubt about these numbers, the load factor of 3% seems quite low. What's your opinion @brunolajoie?
![screenshot at 2018-05-11 09-05-44](https://user-images.githubusercontent.com/7253902/39911122-88a5e422-54fa-11e8-8ee5-6fe3d28550b1.png)

**Comment by maxbellec** [comment]
actually, it seems very unlikely that Nya Öresundsverket in SE-SE4 has a low generation, as it's relatively new (2009).
And Wikipedia says there's another large power plant in SE-SE4, Karlshamn, that has a capacity of 665 MW and burns oil, but this one's quite old.
https://de.wikipedia.org/wiki/Liste_von_Kraftwerken_in_Schweden

Another source is the ENTSOE generation per unit. Most of SE-SE3 is reported (but not everything), and none of SE-SE4 is reported (and Karlshamm is missing).

![screenshot at 2018-05-11 10-14-26](https://user-images.githubusercontent.com/7253902/39914097-432a0cc0-5504-11e8-99f0-b1ee4a31ad44.png)
![screenshot at 2018-05-11 10-14-08](https://user-images.githubusercontent.com/7253902/39914099-444ce172-5504-11e8-8118-cc82c6ab4cf4.png)

**Comment by maxbellec** [comment]
I couldn't find any source of the annual production of any of Karlshamn and Nya Öresundsverket. They could both have a large impact on SE-SE4 carbon intensity, so it'd be great to get some numbers. If anyone can help, that would be great

**Comment by alixunderplatz** [comment]
> "When the Öresund plant was projected, Eon calculated that the municipality of Lund would take its district heating from the plant. This has not happened and Öresundsverket has an overcapacity that Eon uses to produce electricity, which the critics claim increases the emissions. And any change does not seem to be going on."

https://sverigesradio.se/sida/artikel.aspx?programid=96&artikel=5470383

**Comment by maxbellec** [comment]
interesting. They still mention everywhere that it does cogeneration

**Comment by alixunderplatz** [comment]
@maxbellec the link I gave is from 2013, maybe the info is incomplete regarding district heating.

Here are some figures for uniper's assets in Sweden. They report a capacity for gas of 949 MW.
The additional 500 MW come from gas turbines that are distributed over the country (I have not looked into the zones) 

https://www.uniper.energy/sverige/reservkraft/gasturbiner
> The gas turbines are located in Malmö, Barsebäck, Karlshamn and Halmstad and have a combined power of approximately 500 MW."



Total capacity factor for the gas assets in Sweden was around 6% in 2014 and 8.5% in 2015, so your 3% for Öresundsverket seem valid. I guess that it also depends on water levels in the reservoirs and nuclear availability.
Note that the URL structure on uniper's website lists it under "reservkraft", as well as Karlshamn.^^
https://www.uniper.energy/sverige/reservkraft/oresundsverket

This PDF has some info on production:
https://ir.uniper.energy/download/companies/uniperag/VeranstaltungenDownloads/uniper_equity_story_appendix_generation_assets.pdf

![image](https://user-images.githubusercontent.com/25743609/39917075-b8f37baa-550c-11e8-9d58-2d40919bbf03.png)

Karlshamn is only used as reserve/backup power, so no worries about that one. 
One unit was closed after 2014, output was 0 TWh between 2013-2015.
https://www.uniper.energy/sverige/reservkraft/karlshamnsverket

![image](https://user-images.githubusercontent.com/25743609/39917358-8ae254d8-550d-11e8-8cba-99fbf1000035.png)

**Comment by maxbellec** [comment]
The 3% was the load factor of Öresundsverket, by that I meant the total annual generation is 3% of what it would be if it was generating 400MW the whole time, which I don't think is true.

Given there's 440MW in SE-SE4 (Nya Öresundsverket) and 996 in the whole country, does splitting the remaining unknown in half between SE-SE4 and SE-SE3 seem reasonable? @brunolajoie @tmslaine @alixunderplatz?

**Comment by tmslaine** [comment]
While Svenska Kraftnät have not published their annual figures for 2017 as of yet (they are aiming to do it this month) this is what I got from their 2010-2016 annual statistics, units are GWh/year.
[Source](https://www.svk.se/aktorsportalen/elmarknad/statistik/) : Arkiverade dokument, Elområde 3 (SN3) - statistik per månad 2016 etc.

| SE3    | Nuclear    | Wind    | Hydro    | Solar    | GT & diesel    | Other thermal    | Unknown    |
| --- | --- |  --- | --- | --- | --- | --- | --- |
| 2010 | 55777 |  1421 | 11838 | - | 43 | 7321 | 11 |
| 2011 | 58146 |  2726 | 10960 | - | 64 | 6169 | 11 |
| 2012 | 61621 |  3266 | 12928 | 1 | 55 | 5276 | 10 |
| 2013 | 63843 |  4024 | 8851 | 2 | 25 | 5163 | 10 |
| 2014 | 62284 |  4468 | 12511 | 8 | 29 | 4076 | 3 |
| 2015 | 54496 |  6121 | 12539 | 19 | 49 | 4172 | 0 |
| 2016 | 60676 |  5561 | 9260 | 33 | 15 | 4786 | 0 |
| 2017 | - |  - | - | - | - | - | - |

| SE4   | Nuclear    | Wind    | Hydro    | Solar    | GT & diesel    | Other thermal    | Unknown    |
| --- | --- |  --- | --- | --- | --- | --- | --- |
| 2010 | - |  1270 | 1762 | - | 120 | 3892 | 1 |
| 2011 | - |  1879 | 1940 | - | 80 | 2496 | 1 |
| 2012 | - |  2204 | 1853 | - | 33 | 1858 | 1 |
| 2013 | - |  2837 | 1202 | 1 | 14 | 1971 | 2 |
| 2014 | - |  3489 | 1508 | 4 | 5 | 1629 | 2 |
| 2015 | - |  4185 | 1375 | 8 | 5 | 1860 | 2 |
| 2016 | - |  3844 | 1044 | 15 | 4 | 1844 | 2 |
| 2017 | - |  - | - | - | - | - | - |

Grouping Solar, GT & diesel, Other thermal and Unknown together we have

| Unknown  | SE3    | SE4    | SE4 / (SE3+SE4) |
| --- | --- | --- | --- |
| 2010 | 7375 | 4013 | 35.2% |
| 2011 | 6244 | 2577 | 29.2% |
| 2012 | 5342 | 1892 | 26.2% |
| 2013 | 5200 | 1988 | 27.6% |
| 2014 | 4116 | 1640  | 28.4% |
| 2015 | 4240 | 1875  | 30.7% |
| 2016 | 4834 | 1865 | 27.8% |
| 2017 | - | - | - |

Thus, the remaining unknown generation seems to be roughly 70% SE3, 30% SE4.

**Comment by brunolajoie** [comment]
70/30 seems like a reasonable assumption then!

**Comment by maxbellec** [comment]
that more precise @tmslaine, thanks for the investigation. I'll submit in the next days

**Comment by maxbellec** [comment]
As a summary:

## Data we have
- Production data for the whole of Sweden (svk.se)
- Exchanges between sub-zones and between (statnett)
- Consumption (load) in each sub-zone (ENTSOE)
- nuclear production location (100% in SE3)
- wind data in each subzone (ENTSOE)
- 100% of the country's production

## Some stats
- thermal + wind + hydro + nuclear represent 99.93% of the country's production (the rest is Solar and definite Gas)
- In SE1 (resp. SE2), thermal represents 26 / (26 + 1238) = 2.0% (resp 75 / [75 + 2756] = 2.7%) of all non-wind production. Those number are GWh production in 2017.
- In SE3, 6.9% of the production is thermal
- In SE4, 21% of the production is thermal

## "Exact" data we can get
- total production in each subzone (consumption +/- exchange)
- the production data that is not nuclear nor wind is either thermal or hydro

## What we could do and what the margin of error is
- SE1: map 2% of non-wind production to unknown, 98% to hydro. We're probably never more than a few % off here
- SE2: map 3% of non-wind production to unknown, 98% to hydro. We're probably never more than a few % off here 

For SE3 and SE4, two solutions
- make a common SE3-4 area. All nuclear production is there. Map the rest of the country's unknown production to this zone. All non-nuclear, non-wind, non-unknown is hydro. The error will be again a few % max
- separate SE3 and SE4, putting 70% of the country's remaining unkown in SE3, 30% in SE4. If the unkown production is very uncorrelated in SE3 and SE4, and since SE4's production is quite small, we could make up to 30% error on SE4.

**Comment by maxbellec** [comment]
from the summary, if we accept a few percents off, it's safe to split SE into SE1, SE2 and SE3-4. Splitting SE3-4 into SE3 and SE4 could lead to too big errors so we should probably avoid it. Unless we know that thermal production is very correlated in SE3 and SE4.

**Comment by brunolajoie** [comment]
Merging SE3-4 seems a reasonable choice then, until we have an idea of the correlation factor between SE3 and SE4 unknown output. Anyone else has a different opinion on that?

**Comment by Maggus71** [comment]
Just one comment regarding the issue of combining SE3 and SE4, I´m not sure if it´s relevant in this stage, but the big difference between these two bidding areas is that SE3 is more or less in balance when it comes to production and consumption of energy, but in SE4 there is not enough production to cover the consumption. So therefore SE4 needs to import energy on a daily basis to cover the difference. So if you combine the two, it will not really be possible to show the consumer what kind of electricity they are using. I would suggest that you keep them separated.

**Comment by brunolajoie** [comment]
Thanks for the tips @Maggus71. 

It would be nice to split SE3 and SE4 indeed, but:
- In SE3, 6.9% of the production is thermal on average last year
- In SE4, 21% of the production is thermal on average last year
- We can only know in real time the combined sum of SE3 + SE4 thermal production. 

If we split them, we then have to assume that thermal power plants in SE3 and SE4 are activated simultaneously. 

Is that better or worse than grouping the two of them, to your opinion?

**Comment by maxbellec** [comment]
sounds great @Maggus71, we have no hurries here anyway. Better wait and only publish correct data .

**Comment by Maggus71** [comment]
![skarmklipp 2018-05-15 21 34 54](https://user-images.githubusercontent.com/39273556/40079350-e4d5e842-5887-11e8-8a59-2b3d4836a6ab.png)
This is a translation of the installed effect in the four bidding areas. It is very important to understand that this is not the same thing as the production mix in the different areas. The reason is that some installed effect is only used for back up during peak hours (Thermal and gas turbines), the CHP district heating is mainly used during the cold winter season, aprox 15 November-15 Mars. So the production mix in december differs a lot from the production mix in July. The production mix is also effected by the industrys demand for and production of electricity. So for this reason you can not use an production average, you need to set one winter average and one summer average.

**Comment by brunolajoie** [comment]
Thanks for the update! I'm not sure that I was clear enough in my email, but what we'ree trying to do here is finding a way to derive real-time (hourly) production mix per fuel of each Sweden bidding zone, not annual or seasonal averages. 

It seems that we can compute the real time mix for SE1, SE2, and SE3+4 together. Splitting SE3 and SE4 in two will require an assumption over how the "theermal" part of the mix in SE3 and SE4 is split.    If thermal power production in SE3 and SE4 are not correlated at a hourly basis, I feel it's safer not to do the split, and keep SE3+4 grouped

**Comment by magol** [comment]
Is anyone who has sent mail to "Svenska kraftnät" and asked them if they have the data we are missing? I suppose they have access to the data, even if it is not available online. If we are lucky, maybe they can help us.

**Comment by maxbellec** [comment]
it's a good point. I haven't contacted them. @brunolajoie, @tmslaine ?

**Comment by brunolajoie** [comment]
Sent an email to their "contact-me" form. Anyone has better direct contact?

**Comment by Maggus71** [comment]
@brunolajoie I realize that you want to show the realtime production mix for the four bidding areas. My point is that you can not use a yearly average for this since the production mix varies during the year due to the CHP-production which is linked to the demand for district heating. So for example the thermal production in January 2018 lies somewhere in the range of 1250-1600 MWh per day and in July 2017 somewhere in the range of 300-400 MWh per day. So my suggestion is that you use a template for every month built on the historical production data for the last 5-6 years for the production that you do not have any live data of.

**Comment by brunolajoie** [comment]
That's a very interesting idea @Maggus71. It seems technically feasible. My suggestion would be to move on step by step. First, we keep split SE keeping SE3 and S44 together, so we make no assumption of the total level of thermal assumption. And then we open an issue to split SE3 and SE4 together by doing such analysis. What do you guys think?

BTW, I've got an answer from Svenska below
> Dear Bruno!
Thanks for your question and suggestion. Today we do not have the possibility too split production per bidding area. We have a lot of IT-solutions that we have to priority and your suggestion will we take into consideration along with other changes we will do  in the future.

**Comment by Maggus71** [comment]
Here is the installed effect per bidding area in Sweden. Using this you should be able to break down the "other" category and give the users a pretty good picture of what kind of electricity they get in their power outlet.
![skarmklipp 2018-08-13 17 29 06](https://user-images.githubusercontent.com/39273556/44041590-6e1f64e8-9f1e-11e8-8f23-0bd6e2f800ac.png)
![skarmklipp 2018-08-13 17 28 53](https://user-images.githubusercontent.com/39273556/44041591-6e40946a-9f1e-11e8-90ba-f0a5531bf00d.png)
![skarmklipp 2018-08-13 17 28 36](https://user-images.githubusercontent.com/39273556/44041592-6e5f599a-9f1e-11e8-8f13-1842ca5387be.png)

![skarmklipp 2018-08-13 17 28 14](https://user-images.githubusercontent.com/39273556/44041593-6e7dcda8-9f1e-11e8-8cd7-76074b5c92df.png)

**Comment by brunolajoie** [comment]
Super interesting. It will be difficult to implement an winter vs. summer change (eMap is ill equipped to deal with such thing). But your table is helping to make assumptions for the SE3 vs SE4 thermal generation breakdown that we needed. Would you have the same table in GWh instead of MW? That would be even more precise. Thank you!

**Comment by Maggus71** [comment]
I´m sorry I do not have the same figures in GWh. To my knowledge nobody has that since no authority reports the production per bidding area. Since you have the production figures for the wind, hydro and nuclear production for the whole of Sweden already today it would be pretty easy to distribute the production in the four bidding areas according to the installed effect. Then you should be able to distribute the production that is not wind, hydro and nuclear according to the installed effect in four bidding areas. If you want to I can set the installed effect in % instead of MW.

**Comment by corradio** [comment]
Status here:
It seems like there is delayed historical data for bidding zones. However, ENTSOE doesn't report live production data at a sufficiently precise level. We therefore can't split sweden right now.

**Comment by jargonbaron** [comment]
If I may contribute to the discussion here with one link and one proposal.

Production data for all four Swedish bidding zones is published on the Swedish TSO's "extranet" called Mimer. It is close to real-time, but not quite, usually lagging about 12-24 hours behind. So it's not suitable for use in the Electricitymap visualisations, but perhaps the source gets better eventually, and maybe it could help out for historical data?  (.csv donwloads available!)
(https://mimer.svk.se/ProductionConsumption/ProductionIndex, only in Swedish)



The comment from maxbellec (even though it's more than two years old ^^, https://github.com/tmrowco/electricitymap-contrib/issues/1397#issuecomment-388767452) is still a good representation of the current data situation. I understand the philosophy of doing a minimum of calculations and instead relying on accurate source data and not making too many assumptions.

As an alternative proposal, why don't we apply the method of grouping everything that we are not certain of under "unknown", and applying the carbon intensity based on the average generation mix in that bidding zone? Isn't this is the way that the "unknown" production type's emission contribution is currently calculated for Sweden as a whole? I believe a similar thing is done for Netherlands and Croatia for a lack of more accurate data.

Pros:
+ Swedish data can now be published on the bidding zones, allowing also price per zone and installed capacity, among others
+ In my opinion would be the more transparent option (the values are literally Unknown to us), while the carbon intensity value will still be rather accurate
+ This also allows an intuitive solution to issue #2699, as each electricity flow between Sweden and Norway can now be mapped accurately to the individual bidding zones 

Cons:
- Since nearly all production (with the exception of Wind and Nuclear) get grouped under "unknown", we lose all of the nice blue bars associated with Hydro production
- Most inaccurate calculation would be to make the generation mix assumption for SE3 and SE4, as the mix depends highly on season (see earlier comments by Maggus71)
- Still slightly lose accuracy when looking at the country as a whole

Side note on the con no. 2 is that non-wind, non-nuclear part of generation is significant but not big in both SE3 and SE4, as they rely heavily on imports. So when we know Nuclear and Wind and any flows in/out of those two zones, the percentage share that remains unknown is quite small.

I leave this here as a food-for-thought, I know the call to split Sweden has been highly requested but is trickier than it might first seem, and frankly the developers might also be getting tired every time a new issue is opened with the "genius idea" to accurately represent the Swedish bidding zones :)

In any case, I'm prepared to help out if/when any decision is made, and I'm always glad to share my opinion ;-)

![image](https://user-images.githubusercontent.com/26998815/115582249-0e7aff00-a2c9-11eb-8d0c-4b630268e788.png)

**Comment by martincollignon** [comment]
We have been in touch with SVK - I think there is a decent chance they will release the data through ENTSOE soonish :)

**Comment by pierresegonne** [comment]
@Manu1400 has found additional sources for SE in #3247 but mostly related to PRICE or aggregated for the whole country

**Comment by consideRatio** [comment]
@martincollignon do you have any update regarding SVK and them releasing their data through ENTSOE?

Thanks everyone for your effort and work in this amazing project :heart: :tada: :sunflower:!!

**Comment by morganchristiansson** [comment]
As of 15th december ENTSO-E now seems to include all generation in all electricity zones
![image](https://user-images.githubusercontent.com/66844/146586180-78a6a2d5-4127-4b85-abbc-74750a1382de.png)

Does this mean we have all data and can complete this task?

**Comment by martian-maniac** [comment]
As of 15th december ENTSO-E now seems to include all generation in all electricity zones ![image](https://user-images.githubusercontent.com/66844/146586180-78a6a2d5-4127-4b85-abbc-74750a1382de.png)

[source](https://transparency.entsoe.eu/generation/r2/actualGenerationPerProductionType/show?name=&defaultValue=false&viewType=GRAPH&areaType=BZN&atch=false&datepicker-day-offset-select-dv-date-from_input=D&dateTime.dateTime=15.12.2021+00:00%7CUTC%7CDAYTIMERANGE&dateTime.endDateTime=15.12.2021+00:00%7CUTC%7CDAYTIMERANGE&area.values=CTY%7C10YSE-1--------K!BZN%7C10Y1001A1001A44P&productionType.values=B01&productionType.values=B02&productionType.values=B03&productionType.values=B04&productionType.values=B05&productionType.values=B06&productionType.values=B07&productionType.values=B08&productionType.values=B09&productionType.values=B10&productionType.values=B11&productionType.values=B12&productionType.values=B13&productionType.values=B14&productionType.values=B20&productionType.values=B15&productionType.values=B16&productionType.values=B17&productionType.values=B18&productionType.values=B19&dateTime.timezone=UTC&dateTime.timezone_input=UTC)

I believe ENTSOE.py is already fetching all generation data (was previously wind only). So the data is already parsed and no parser update required I think? Does this mean we have all data and can complete this task?

**Comment by morganchristiansson** [comment]
[ENTSOE.query_production()](https://github.com/electricityMap/electricitymap-contrib/blob/4fe7483c26b66cd47b44e5b2143086df72aee5da/parsers/ENTSOE.py#L444-L450) should already be returning the new metrics and datapoints into your systems...

We should stop using SE.py and replace with data from ENTSOE.py.. 

I suspect it's being called from outside of this `-contrib` repo? Is it possible to contribute?

**Comment by martian-maniac** [comment]
I discussed this @corradio on Twitter and there is data issue with Hydro generation reported by ENTSO-E vs SvK (SE.py).

I created below graph comparing SE Hydro generation from ENTSOE and SvK. As of 15th dec ENTSO-E is reporting up to 7GW / 50% higher values than SvK.

I suspect ENTSO-E is incorrect - it reported 15GW peak output from Hydro on the 17th dec. This would be a new record, SE Hydro has never gone above 13GW before.

This is likely a blocker to delivering this task. Altho work can progress meanwhile, the data is there just not correct (yet).

![image](https://user-images.githubusercontent.com/96312338/146695292-44b4a93b-04f2-45ce-9a97-b6ef6c2f3b4f.png)

Other than this data issue the data matches and looks good!

**Comment by martian-maniac** [comment]
As a workaround we could use the ratios from ENTSO-E SE1/2/3/4 and correct it with total from current source in SE.py.

I.e. SE1/2/3/4= 10GW/5GW/0/0 = total 15GW
total from current source=10GW= 1.5x too high
adjusted values SE1/2/3/4= 6.6GW/3.3GW/0/0 = total 10GW

I've not received any acknowledgement of the hydro data issue. Hydro generation continues to be overreported by up to ~50% exceeding the installed capacity for the country.

**Comment by gwpicard** [comment]
I just got a message from SVK yesterday -- according to them, the data has been updated on ENSTO-E going forward for waste, solar, other renewables, other, nuclear and hydro. 

Would someone be able to run an investigation on this new data coming in to see if indeed the quality has improved? It would be so nice to finally have the Swedish subzones.

**Comment by martian-maniac** [comment]
Here is updated graph I posted earlier. Hydro continues to be overreported in ENTSO-E data and is frequently over installed hydro capacity of 16GW

![image](https://user-images.githubusercontent.com/66844/148933262-95e3f060-0bf7-4a1d-9814-9e01f532be1d.png)

There is some minor difference in ENTSO-E datapoints for other production types but I think it's within margin of error.. It matches very closely.

![image](https://user-images.githubusercontent.com/96312338/148933700-daa089ab-a381-4992-a215-8f9f8bc9aebe.png)

**Comment by consideRatio** [comment]
@martian-maniac you have a very clear and distinct bug observed in the hydro data reported to ENTSO-E. I presume there is a mistake in the reporting from SVK to ENTSO-E. My guess is that there is some double-reporting of some of the hydro production.

Have you or someone else reported this to SVK already?

@gwpicard for reference so others can help out with issues like this in the future, how did you contact SVK? Did you email a certain email-address or use some web form etc?

**Comment by gwpicard** [comment]
Thanks for the analysis @martian-maniac. I think perhaps the data is being updated only recently in the last few days, I can confirm with them.

@consideRatio someone from SVK reached out to us directly about this.

**Comment by martian-maniac** [comment]
Still ongoing...
![image](https://user-images.githubusercontent.com/96312338/149708340-18a20e9f-40a1-4e0c-939d-19613a1fb7e7.png)

**Comment by gwpicard** [comment]
@martian-maniac thanks for the ongoing analysis. I've sent it to the contact who reached out from @SVK, and he's now investigating. He'll reach out to me with updates when he has them

**Comment by martian-maniac** [comment]
One more week passed. Still ongoing...
![image](https://user-images.githubusercontent.com/96312338/150641683-ef9f665e-4ce4-4fe9-bcc2-a7d4842ddf94.png)

**Comment by gwpicard** [comment]
@martian-maniac Good news: the person I'm in touch with at SVK just told me that they've identified the root issue and they're now working on it. 
It will take them a bit of time to fix because of it involves various pieces I'll keep you posted once he lets me know that they've fixed it

**Comment by consideRatio** [comment]
@gwpicard it would be great to have people from SvK active on GitHub, I hope you encourage your connections to take the step to interact directly here :)

**Comment by gwpicard** [comment]
@consideRatio Agreed! It's not always easy 😆

**Comment by martian-maniac** [comment]
As of 25th January the data issue is resolved. (i.e. day after root cause was confirmed here).
Sorry for delay in analysis, I've been away and just regained access to my homelab.

So we have full and correct data for SE1/2/3/4 from ENTSO-E now.. Should be good to switch from current SE.py parser to ENTSOE.py

![image](https://user-images.githubusercontent.com/96312338/154026275-9534f4f0-ccfe-471d-afe8-fb3dc5541e7b.png)

**Comment by gwpicard** [comment]
@martian-maniac this is great! And no worries at all, thanks for following this so closely. I'll circle back in a bit on this once I've touched base with SvK. And to your point @consideRatio, they aren't currently on Github...

**Comment by gwpicard** [comment]
An update on the status here - looks like the data is a go from SvK. Unfortunately, we can't switch over to the subzone data because currently we can't test out regional parsers without removing the national ones. We still have people using the national data and we don't want to remove this access. We know other regions like IT, DK or NO are split, but these changes were made much before. We have a similar issue with FR which is blocking until we have a simple way to migrate from country data to regional data.

For now, we will keep this open until we can merge once we've made an update to our infrastructure

**Comment by magol** [comment]
@gwpicard Is this update to the infrastructure connected to a issue and what time frame is we talking about?

**Comment by gwpicard** [comment]
@magol There isn't a specific issue on this at the moment, though it was previously mentioned for example here: https://github.com/electricitymap/electricitymap-contrib/pull/2613#issuecomment-785899959

The time frame is difficult to say at the moment, because it would involve some changes in our backend infrastructure which we haven't shaped yet.

**Comment by morganchristiansson** [comment]
Could the national data just be hidden but still available to users (I assume API users)?

**Comment by martian-maniac** [comment]
Could you calculate national average Co2 from multiple regions? SE = SE1+2+3+4 / 4 (weighted by consumption).

Then you don't need to flow-trace in parallel.

Though SE.py has per-minute datapoints from SvK controlroom api tho while ENTSO-E is hourly. Not sure if this is important. Larger step changes appear to happen hourly anyway.

**Comment by gwpicard** [comment]
Indeed we could do that, but we would prefer having both geographical levels accessible (and flow-traced) rather than create a need for a manual calculation. This is also needed to be able to release for different geographies as they become available.

In terms of the SE per-minute datapoints from SvK, at the minute we only compute hourly so it won't be relevant unfortunately for a little while until we can increase temporal granularity...

**Comment by magol** [comment]
@pierresegonne @VIKTORVAV99 
Now one hour after the merge, I see on the live site that Sweden is splitet up, but it don't show any data. Is this temporary, or is it a problem?

![image](https://user-images.githubusercontent.com/493140/187886851-d021b956-da66-4926-a6a8-05162ff6158e.png)

**Comment by pierresegonne** [comment]
Our data processing pipeline is not 100% real-time but data should appear soon :)

**Comment by VIKTORVAV99** [comment]
> Our data processing pipeline is not 100% real-time but data should appear soon :)

Did you guys pull historical data as well? Should be available since 2021-12-15 for the individual bidding zones.

**Comment by VIKTORVAV99** [comment]
@magol data should be available now! 🎉

I noticed the exchange arrows are misaligned though but I'll take care of that later.

**Comment by martian-maniac** [comment]
DK1 (West Denmark) -> SE3 (South Central Sweden) arrow is incorrectly pointing at SE4 (South Sweden). All data looks correct. I assume this is what @VIKTORVAV99 meant by misaligned exchange arrows.

![image](https://user-images.githubusercontent.com/96312338/188328256-433e4c32-7fa3-40af-aad4-a381a04d7ac4.png)

**Comment by VIKTORVAV99** [comment]
> DK1 (West Denmark) -> SE3 (South Central Sweden) arrow is incorrectly pointing at SE4 (South Sweden). All data looks correct. I assume this is what @VIKTORVAV99 meant by misaligned exchange arrows.
> 
> ![image](https://user-images.githubusercontent.com/96312338/188328256-433e4c32-7fa3-40af-aad4-a381a04d7ac4.png)
> 

Yep, that whole border between SE3 and SE4 is a bit further up than it should be but that's a later issue.

I'll move the exchange arrows as soon as the mock data is updated and they are visible in local development (it's a pain in the ass otherwise).

I was gonna ask the team about it when they are back in the office tomorrow (Monday) since they are the only ones that can generate the mock data.

**Comment by monsieurhannes** [comment]
Is there any way to find the live "output" of Karlshamns oil power plant in SE4? It hasn't been very active in recent years but it got started already in summer this year as it became economical due to the high prices. 
Currently there is some "unknown" power in SE4 (246 MWh), which could include Karlshamn, with a roughly 3x higher carbon intensity than "unknown".
svk.se doesn't list the different power plants individually...
![image](https://user-images.githubusercontent.com/20651440/188499924-a52efad6-88da-4376-ab7d-ea596396aab6.png)

**Comment by VIKTORVAV99** [comment]
> Is there any way to find the live "output" of Karlshamns oil power plant in SE4? It hasn't been very active in recent years but it got started already in summer this year as it became economical due to the high prices. 
> Currently there is some "unknown" power in SE4 (246 MWh), which could include Karlshamn, with a roughly 3x higher carbon intensity than "unknown".
> svk.se doesn't list the different power plants individually...
> ![image](https://user-images.githubusercontent.com/20651440/188499924-a52efad6-88da-4376-ab7d-ea596396aab6.png)
> 
> 

Sadly not live data, as far as I have been able to tell. The data in ENTSO-E does not split out oil (yet at least) and the data available for the plant specific data is almost a week old.

And if I'm not mistaken the ENTSO-E API do not return production per generation unit until Transparency Platform version 3.02 is released. (That should be tomorrow if everything goes according to plan). 

@pierresegonne maybe you could shed some light on how the generation per unit is used and if we could incorporate this into the historical data for SE-SE4? (it should replace an equal amount of unknown production)

**Comment by pierresegonne** [comment]
> @pierresegonne maybe you could shed some light on how the generation per unit is used and if we could incorporate this into the historical data for SE-SE4? (it should replace an equal amount of unknown production)

It's not 😅 we parse it but currently only use production events

**Comment by VIKTORVAV99** [comment]
@martian-maniac I think the best bet for getting this "live" data would be to reach out to SVK and push them towards including it in the ENTSO-E transparency platform.

I'll eventually update the unknown CO2 intensity data when the new figures are published but right now they are an average of Sweden as a whole.

Hopefully we can find a solution to calculate it more accurately per zone soon.

**Comment by pierresegonne** [comment]
Good idea! I think they've been pretty responsive in the past. @gwpicard do you think the contact we have could help us ask for opening up more data?

**Comment by gwpicard** [comment]
I've just shot them a message — unfortunately they're not on GitHub so I can't ping them here. Will keep everyone posted.
